File: cpp/sai_test.cc
// Copyright 2013 The CARFAC Authors. All Rights Reserved.
// Author: Ron Weiss <ronw@google.com>
//
// This file is part of an implementation of Lyon's cochlear model:
// "Cascade of Asymmetric Resonators with Fast-Acting Compression"
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "sai.h"

#include <cmath>
#include <iostream>
#include <string>

#include "gtest/gtest.h"

#include "agc.h"
#include "car.h"
#include "carfac.h"
#include "common.h"
#include "ihc.h"
#include "test_util.h"
#include <Eigen/Core>

using testing::Values;

class SAIPeriodicInputTest
    : public SAITestBase,
      public ::testing::WithParamInterface< ::testing::tuple<int, int> > {
 protected:
  void SetUp() {
    period_ = ::testing::get<0>(GetParam());
    num_channels_ = ::testing::get<1>(GetParam());
  }

  int period_;
  int num_channels_;
};

TEST_P(SAIPeriodicInputTest, MultiChannelPulseTrain) {
  const int kInputSegmentWidth = 38;
  ArrayXX segment =
      CreatePulseTrain(num_channels_, kInputSegmentWidth, period_);

  const int kSAIWidth = 15;
  const int kTriggerWindowWidth = kInputSegmentWidth;
  SAIParams sai_params = CreateSAIParams(num_channels_, kInputSegmentWidth,
                                         kTriggerWindowWidth, kSAIWidth);
  sai_params.future_lags = 0;  // Only compute past lags.

  SAI sai(sai_params);
  ArrayXX sai_frame;
  sai.RunSegment(segment, &sai_frame);

  // The output should have peaks at the same positions, regardless of
  // input phase.
  for (int i = 0; i < num_channels_; ++i) {
    const ArrayX& sai_channel = sai_frame.row(i);
    for (int j = sai_channel.size() - 1; j >= 0; j -= period_) {
      EXPECT_TRUE(HasPeakAt(sai_channel, j));
    }
  }

  std::cout << "Input:" << std::endl << segment << std::endl
            << "Output:" << std::endl << sai_frame << std::endl;
}

INSTANTIATE_TEST_SUITE_P(PeriodicInputVariations, SAIPeriodicInputTest,
                         testing::Combine(Values(25, 10, 5, 2),  // periods.
                                          Values(1, 2, 15)));  // num_channels.

class SAITest : public SAITestBase {};

TEST_F(SAITest, DiesIfInputSegmentWidthIsLargerThanBuffer) {
  const int kNumChannels = 2;
  const int kSAIWidth = 10;
  const int kInputSegmentWidth = 200;
  SAIParams valid_params = CreateSAIParams(kNumChannels, kInputSegmentWidth,
                                           kInputSegmentWidth, kSAIWidth);
  SAI sai(valid_params);

  const int kTriggerWindowWidth = kInputSegmentWidth / 10;
  SAIParams params = CreateSAIParams(kNumChannels, kInputSegmentWidth,
                                     kTriggerWindowWidth, kSAIWidth);
  ASSERT_GT(kInputSegmentWidth,
            params.num_triggers_per_frame * params.trigger_window_width);
  ASSERT_DEATH(sai.Redesign(params), "Too few triggers");
}

TEST_F(SAITest, DiesIfInputWidthDoesntMatchInputSegmentWidth) {
  const int kNumChannels = 2;
  const int kInputSegmentWidth = 10;
  const int kPeriod = 4;
  ArrayXX segment = CreatePulseTrain(kNumChannels, kInputSegmentWidth, kPeriod);

  const int kSAIWidth = 20;
  const int kExpectedInputSegmentWidth = kInputSegmentWidth - 1;
  const int kTriggerWindowWidth = kSAIWidth + 1;
  SAIParams sai_params = CreateSAIParams(
      kNumChannels, kExpectedInputSegmentWidth, kTriggerWindowWidth, kSAIWidth);
  ASSERT_NE(sai_params.input_segment_width, kInputSegmentWidth);
  SAI sai(sai_params);
  ArrayXX sai_frame;
  ASSERT_DEATH(sai.RunSegment(segment, &sai_frame), "input samples");
}

TEST_F(SAITest, InputSegmentWidthSmallerThanTriggerWindow) {  // I.e. small hop.
  const int kNumChannels = 1;
  const int kTotalInputSamples = 20;
  const int kPeriod = 5;
  ArrayXX input = CreatePulseTrain(kNumChannels, kTotalInputSamples, kPeriod);

  const int kNumFrames = 4;
  const int kInputSegmentWidth = kTotalInputSamples / kNumFrames;
  const int kTriggerWindowWidth = kTotalInputSamples;
  const int kSAIWidth = 15;
  SAIParams sai_params = CreateSAIParams(kNumChannels, kInputSegmentWidth,
                                         kTriggerWindowWidth, kSAIWidth);
  ASSERT_LT(sai_params.input_segment_width, sai_params.trigger_window_width);
  sai_params.future_lags = 0;  // Only compute past lags.

  ASSERT_GE(kPeriod, kInputSegmentWidth);

  SAI sai(sai_params);
  ArrayXX sai_frame;
  for (int i = 0; i < kNumFrames; ++i) {
    ArrayXX segment =
        input.row(0).segment(i * kInputSegmentWidth, kInputSegmentWidth);
    sai.RunSegment(segment, &sai_frame);

    std::cout << "Frame " << i << std::endl
              << "Input:" << std::endl << segment << std::endl
              << "Output:" << std::endl << sai_frame << std::endl;

    EXPECT_NE(segment.cwiseAbs().sum(), 0);
    // Since the input segment is never all zero, there should always
    // be a peak at zero lag.
    ArrayX sai_channel = sai_frame.row(0);
    EXPECT_TRUE(HasPeakAt(sai_channel, sai_channel.size() - 1));

    if (i == 0) {
      // Since the pulse train period is larger than the input segment
      // size, the first input segment will only see a single impulse,
      // most of the SAI will be zero.
      AssertArrayNear(sai_channel.head(sai_channel.size() - 1),
                      ArrayX::Zero(sai_channel.size() - 1), 1e-9);
    }

    if (i == kNumFrames - 1) {
      // By the last frame, the SAI's internal buffer will have
      // accumulated the full input signal, so the resulting image
      // should contain kPeriod peaks.
      for (int j = sai_channel.size() - 1; j >= 0; j -= kPeriod) {
        EXPECT_TRUE(HasPeakAt(sai_channel, j));
      }
    }
  }
}

TEST_F(SAITest, MatchesMatlabOnBinauralData) {
  const std::string kTestName = "binaural_test";
  const int kInputSegmentWidth = 882;
  const int kNumChannels = 71;
  // The Matlab CARFAC output is transposed compared to the C++.
  ArrayXX input_segment =
      LoadMatrix(kTestName + "-matlab-nap1.txt", kInputSegmentWidth,
                 kNumChannels).transpose();

  const int kTriggerWindowWidth = kInputSegmentWidth;
  const int kSAIWidth = 500;
  SAIParams sai_params = CreateSAIParams(kNumChannels, kInputSegmentWidth,
                                         kTriggerWindowWidth, kSAIWidth);
  SAI sai(sai_params);
  ArrayXX sai_frame;
  sai.RunSegment(input_segment, &sai_frame);
  ArrayXX expected_sai_frame = LoadMatrix(kTestName + "-matlab-sai1.txt",
                                          kNumChannels, sai_params.sai_width);
  AssertArrayNear(expected_sai_frame, sai_frame, kTestPrecision);

  WriteMatrix(kTestName + "-cpp-sai1.txt", sai_frame);
}

TEST_F(SAITest, CARFACIntegration) {
  const int kNumEars = 1;
  const int kInputSegmentWidth = 300;
  ArrayXX segment(kNumEars, kInputSegmentWidth);

  // Sinusoid input.
  const float kFrequency = 10.0;
  segment.row(0) =
      ArrayX::LinSpaced(kInputSegmentWidth, 0.0, 2 * kFrequency * M_PI).sin();

  CARParams car_params;
  IHCParams ihc_params;
  AGCParams agc_params;
  CARFAC carfac(kNumEars, 8000, car_params, ihc_params, agc_params);
  CARFACOutput output(true, false, false, false);
  const bool kOpenLoop = false;
  carfac.RunSegment(segment, kOpenLoop, &output);

  const int kSAIWidth = 20;
  const int kTriggerWindowWidth = kInputSegmentWidth;
  SAIParams sai_params =
      CreateSAIParams(carfac.num_channels(), kInputSegmentWidth,
                      kTriggerWindowWidth, kSAIWidth);
  SAI sai(sai_params);
  ArrayXX sai_frame;
  sai.RunSegment(output.nap()[0], &sai_frame);

  // TODO(ronw): Test something about the output.
}

File: python/sai_test.py
# Lint as: python3
#!/usr/bin/env python

# Copyright 2013 The CARFAC Authors. All Rights Reserved.
#
# This file is part of an implementation of Lyon's cochlear model:
# "Cascade of Asymmetric Resonators with Fast-Acting Compression"
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http:#www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Tests for carfac.python.sai."""

import os
import unittest

import numpy as np

import sai as pysai

_TEST_DATA_DIR = "../test_data"


def LoadMatrix(filename, rows, columns):
  """Reads a matrix with shape (rows, columns) matrix from a text file."""
  matrix = np.loadtxt(os.path.join(_TEST_DATA_DIR, filename))
  assert matrix.shape == (rows, columns)
  return matrix


def WriteMatrix(filename, matrix):
  np.savetxt(os.path.join(_TEST_DATA_DIR, filename), matrix, fmt="%0.12f")


def CreatePulseTrain(num_channels, num_samples, period, leading_zeros=0):
  segment = np.zeros((num_channels, num_samples))
  for i in range(num_channels):
    # Begin each channel at a different phase.
    phase = (i + leading_zeros) % period
    for j in range(phase, num_samples, period):
      segment[i, j] = 1
  return segment


def CreateSAIParams(sai_width, num_triggers_per_frame=2, **kwargs):
  """Fills an SAIParams object using reasonable defaults for some fields."""
  return pysai.SAIParams(sai_width=sai_width,
                         # Half of the SAI should come from the future.
                         future_lags=sai_width // 2,
                         num_triggers_per_frame=num_triggers_per_frame,
                         **kwargs)


def HasPeakAt(frame, index):
  if index == 0:
    return frame[index] > frame[index + 1]
  elif index == len(frame) - 1:
    return frame[index] > frame[index - 1]
  return frame[index] > frame[index + 1] and frame[index] > frame[index - 1]


class PeriodicInputTest(unittest.TestCase):
  def _RunMultiChannelPulseTrainTest(self, period, num_channels):
    input_segment_width = 38
    segment = CreatePulseTrain(num_channels, input_segment_width, period)

    sai_params = CreateSAIParams(num_channels=num_channels,
                                 input_segment_width=input_segment_width,
                                 trigger_window_width=input_segment_width,
                                 sai_width=15)
    sai_params.future_lags = 0  # Only compute past lags.

    sai = pysai.SAI(sai_params)
    sai_frame = sai.RunSegment(segment)

    # The output should have peaks at the same positions, regardless of
    # input phase.
    for i in range(num_channels):
      sai_channel = sai_frame[i, :]
      for j in range(len(sai_channel) - 1, 0, -period):
        print(i, j, sai_channel, HasPeakAt(sai_channel, j))
        self.assertTrue(HasPeakAt(sai_channel, j))

    print("Input\n{}\nOutput\n{}".format(segment, sai_frame))

  def testMultiChannelPulseTrain(self):
    for period in [25, 10, 5, 2]:
      for num_channels in [1, 2, 15]:
        print("Testing period={}, num_channels={}".format(period, num_channels))
        self._RunMultiChannelPulseTrainTest(period, num_channels)


class SAITest(unittest.TestCase):
  def testInputSegmentWidthIsLargerThanBuffer(self):
    params = CreateSAIParams(num_channels=2, sai_width=10,
                             input_segment_width=200,
                             trigger_window_width=200)
    sai = pysai.SAI(params)

    params.trigger_window_width = params.input_segment_width // 10
    self.assertGreater(params.input_segment_width,
                       params.num_triggers_per_frame *
                       params.trigger_window_width)
    self.assertRaises(AssertionError, sai.Redesign, params)

  def testInputWidthDoesntMatchInputSegmentWidth(self):
    num_channels = 2
    input_segment_width = 10
    segment = CreatePulseTrain(num_channels, input_segment_width, period=4)

    sai_width = 20
    expected_input_segment_width = input_segment_width - 1
    sai_params = CreateSAIParams(
        num_channels=num_channels,
        sai_width=sai_width,
        input_segment_width=expected_input_segment_width,
        trigger_window_width=sai_width + 1)
    self.assertNotEqual(sai_params.input_segment_width, input_segment_width)
    sai = pysai.SAI(sai_params)
    self.assertRaises(AssertionError, sai.RunSegment, segment)

  def testInputSegmentWidthSmallerThanTriggerWindow(self):
    """Tests small hop between segments."""
    num_channels = 1
    total_input_samples = 20
    period = 5
    full_input = CreatePulseTrain(num_channels, total_input_samples, period)

    num_frames = 4
    input_segment_width = total_input_samples // num_frames
    sai_params = CreateSAIParams(num_channels=num_channels,
                                 input_segment_width=input_segment_width,
                                 trigger_window_width=total_input_samples,
                                 sai_width=15)
    self.assertLess(sai_params.input_segment_width,
                    sai_params.trigger_window_width)
    sai_params.future_lags = 0  # Only compute past lags.

    self.assertGreaterEqual(period, input_segment_width)

    sai = pysai.SAI(sai_params)
    for i in range(num_frames):
      segment = (
          full_input[:, i * input_segment_width:(i+1) * input_segment_width])
      sai_frame = sai.RunSegment(segment)

      print("Frame {}\nInput\n{}\nOutput\n{}".format(i, segment, sai_frame))

      self.assertNotEqual(np.abs(segment).sum(), 0)
      # Since the input segment is never all zero, there should always
      # be a peak at zero lag.
      sai_channel = sai_frame[0, :]
      self.assertTrue(HasPeakAt(sai_channel, len(sai_channel) - 1))

      if i == 0:
        # Since the pulse train period is larger than the input segment
        # size, the first input segment will only see a single impulse,
        # most of the SAI will be zero.
        np.testing.assert_allclose(sai_channel[:len(sai_channel) - 1],
                                   np.zeros(len(sai_channel) - 1), 1e-9)

      if i == num_frames - 1:
        # By the last frame, the SAI's internal buffer will have
        # accumulated the full input signal, so the resulting image
        # should contain kPeriod peaks.
        for j in range(len(sai_channel) - 1, 0, -period):
          self.assertTrue(HasPeakAt(sai_channel, j))

  def testMatchesMatlabOnBinauralData(self):
    test_name = "binaural_test"
    input_segment_width = 882
    num_channels = 71
    # The Matlab CARFAC output is transposed compared to the C++.
    input_segment = LoadMatrix(test_name + "-matlab-nap1.txt",
                               input_segment_width,
                               num_channels).transpose()

    sai_params = CreateSAIParams(num_channels=num_channels,
                                 input_segment_width=input_segment_width,
                                 trigger_window_width=input_segment_width,
                                 sai_width=500)
    sai = pysai.SAI(sai_params)
    sai_frame = sai.RunSegment(input_segment)
    expected_sai_frame = LoadMatrix(test_name + "-matlab-sai1.txt",
                                    num_channels, sai_params.sai_width)
    np.testing.assert_allclose(expected_sai_frame, sai_frame, rtol=1e-5)

    WriteMatrix(test_name + "-py-sai1.txt", sai_frame)


if __name__ == "__main__":
  unittest.main()
